2016/presentations/object_calisthenics/text.md

```diff
@@ -0,0 +1,94 @@
+
+ Zupełnym przeciwieństwem jest kilku bądź kilkunastolinijkowa metoda, która ma wyłącznie jedną odpowiedzialność - robi jedną rzecz, robi ją dobrze, a dodatkowo jest adekwatnie nazwana. Dobra nazwa metody może znacznie ułatwić zrozumienie jej celu w trakcie czytania kodu. Jeśli wiemy, co dokładnie jest celem metody, to chętniej jej użyjemy w innym miejscu. Ponadto krótki kod, który zmienia stan wyłącznie jednego obiektu i nie posiada efektów ubocznych, jest łatwo testowalny.
+
+## 2. Nie używaj słowa kluczowego `else`
+
+ Konstrukcja `if/else` jest obecna w prawie każdym języku programowania i łatwo ją zrozumie nawet laik. Wielu z nas zna przypadek wielokrotnie powtarzanych ciągów `if … elif … elif … else`. Trudno się je czyta, trudno zrozumieć, częściej trudno zmodyfikować - a pokusa dopisania kolejnego `else` jest ogromna. Tego typu sytuacje często prowadzą do duplikowania kodu oraz wielu błędów.
+
+ Jednym z rozwiązań pozwalającym unikać takich konstrukcji jest polimofizm, wzorzec strategii lub wzorzec stanu. Kod, który korzysta z takiego podejścia, znacznie łatwiej się czyta i jest łatwiejszy w utrzymaniu.
+
+ W najprostszym przypadku wystarczy, że sprawdzamy warunek w `if` i wychodzimy jeśli nie został on spełniony. Dzięki temu nie mamy w kodzie kilku możliwych rozgałęzień. Reguła ta jest prostsza do przestrzegania, jeśli każda metoda robi tylko jedną rzecz.
+
+## 3. Prymitywne typy powinny być opakowane jako klasy.
+
+ Powyższa zasada obowiązuje wyłącznie w sytuacji, gdy zmienna przekazuje jakieś zachowanie. Do czasu Pythona 3.5 nie było możliwości by wymusić na poziomie składni by metoda przyjmowała konkretny typ zmiennej. Nawet, jeśli mamy dostępne tego typu udogodnienie, powyższa zasada ma sens.
+
+ Obowiązek poinformowania o intencjach metody spoczywa niejako na nazwie metody. Jeśli zamiast integer przekażemy do klasy obiekt klasy Hour lub Year już sama deklaracja metody jest czytelniejsza. Nie przekażemy przez pomyłkę obiektu Year, gdy metoda spodziewa się od nas Hour lub odwrotnie - co jest możliwe przy samych typach prostych.
+
+ Tym samym dajemy programiście proste i czytelne API, które może nie wymagać nawet wgłębiania się w kod implementacji. Co więcej, jeśli już mamy tego typu klasy opakowujące typy proste, to jest to naturalne miejsce by dodać kolejne metody operujące na tych typach.
+
+## 4. Tylko jedna kropka na linię.
+
+ (nie dotyczy getterów i tzw. "Fluent interface")
+ Celem tej zasady jest powtrzymanie programisty przed sięganie zbyt głęboko w implementację klas i tym samym łamanie enkapsulacji. Jeśli w ramach jednej linii odwołujemy się do dwóch różnych obiektów, to znaczy że wiemy zbyt wiele o innych obiektach.
```

Comment:
"powstrzymanie programisty przed sięganie" -- chyba "przed sięganiem"? to samo nie "łamanie" tylko "łamaniem".